spec/formatting.md

```diff
@@ -57,6 +57,25 @@ nor be made available to _function handlers_.
 > _declarations_ affecting _variables_ referenced by that _expression_
 > have already been evaluated in the order in which the relevant _declarations_
 > appear in the _message_.
+> An implementation MUST ensure that every _expression_ in a _message_
+> is evaluated at most once.
+>
+> Implementations and users SHOULD NOT create _function handlers_
+> that mutate external program state,
+> particularly since such a _function handler_ can present a remote execution hazard.
+>
+
+> [!NOTE]
+>
+> Function handlers can be written that depend on external mutable state
+> (for example, the current date and time).
+> Lazy evaluation might involve evaluating the same _expression_ multiple times
+> (call-by-name) or evaluating every expression at most once (call-by-need).
+> This justifies the requirement to ensure that every _expression_ is
+> evaluated at most once: different invocations of the same function handler
+> might return different results (functions are not pure).
+> This means that lazy implementations cannot use
+> a call-by-name evaluation strategy.
```

Comment:
I think this note is burying the lede. It should appear directly after the (first) "IMPORTANT", where we forbid multiple evaluations and I think it should start with the guidance. Something like:

> [!NOTE]
> Implementations that choose lazy evaluation cannot use a call-by-name
> evaluation strategy (instead of evaluating expressions at most once, or 
> "call-by-need").
> This is to prevent _function handlers_ that depend on external mutable state
> (for example, ones that obtain the current system clock time)
> from returning different values for the same _expression_ in different
> parts of a given _message_.

Reply:
Took another pass at it in [fdbb48c](https://github.com/unicode-org/message-format-wg/pull/901/commits/fdbb48c57ca804e652902bb05951a98f64d29f6f) -- let me know what you think.